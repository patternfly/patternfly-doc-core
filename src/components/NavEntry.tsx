import { NavItem } from '@patternfly/react-core'

export interface TextContentEntry {
  id: string
  data: {
    id: string
    section: string
  }
  collection: string
}

interface NavEntryProps {
  entry: TextContentEntry
  isActive: boolean
}

export const NavEntry = ({ entry, isActive }: NavEntryProps) => {
  const { id } = entry
  const { id: entryTitle, section } = entry.data

  return (
    <NavItem itemId={id} to={`/${section}/${id}`} isActive={isActive} id={`nav-entry-${id}`}>
      {entryTitle}
    </NavItem>
  )
}
