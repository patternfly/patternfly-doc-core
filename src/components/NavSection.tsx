import { NavExpandable } from '@patternfly/react-core'
import { sentenceCase } from 'change-case'
import { NavEntry, type TextContentEntry } from './NavEntry'

interface NavSectionProps {
  entries: TextContentEntry[]
  sectionId: string
  activeItem: string
}

export const NavSection = ({
  entries,
  sectionId,
  activeItem,
}: NavSectionProps) => {
  const isExpanded = window.location.pathname.includes(sectionId)

  const sortedNavEntries = entries.sort((a, b) =>
    a.data.id.localeCompare(b.data.id),
  )

  const isActive = sortedNavEntries.some((entry) => entry.id === activeItem)

  const items = sortedNavEntries.map((entry) => (
    <NavEntry key={entry.id} entry={entry} isActive={activeItem === entry.id} />
  ))

  return (
    <NavExpandable
      title={sentenceCase(sectionId)}
      isActive={isActive}
      isExpanded={isExpanded}
      id={`nav-section-${sectionId}`}
    >
      {items}
    </NavExpandable>
  )
}
