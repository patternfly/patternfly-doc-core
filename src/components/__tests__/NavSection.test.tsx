import { render, screen } from '@testing-library/react'
import { NavSection } from '../NavSection'
import { type TextContentEntry } from '../NavEntry'

const mockEntries: TextContentEntry[] = [
  {
    id: 'entry1',
    data: { id: 'Entry1', section: 'section1' },
  },
  {
    id: 'entry2',
    data: { id: 'Entry2', section: 'Section1' },
  },
  {
    id: 'entry3',
    data: { id: 'Entry3', section: 'Section1' },
  },
]

it('renders without crashing', () => {
  render(
    <NavSection
      entries={mockEntries}
      sectionId="section1"
      activeItem="entry1"
    />,
  )
  expect(screen.getByText('Section1')).toBeInTheDocument()
})

it('collapses if the sectionId is not in the pathname', () => {
  Object.defineProperty(window, 'location', {
    value: {
      pathname: '/foo',
    },
    writable: true,
  })

  render(
    <NavSection
      entries={mockEntries}
      sectionId="section1"
      activeItem="entry1"
    />,
  )

  expect(screen.getByText('Section1')).toHaveAttribute('aria-expanded', 'false')
})

it('expands if the sectionId is in the pathname', () => {
  Object.defineProperty(window, 'location', {
    value: {
      pathname: '/foo/section1',
    },
    writable: true,
  })

  render(
    <NavSection
      entries={mockEntries}
      sectionId="section1"
      activeItem="entry1"
    />,
  )

  expect(screen.getByText('Section1')).toHaveAttribute('aria-expanded', 'true')
})

it('renders the correct number of entries', () => {
  render(
    <NavSection
      entries={mockEntries}
      sectionId="section1"
      activeItem="entry1"
    />,
  )

  expect(screen.getAllByRole('listitem')).toHaveLength(4)
})

it('renders all passed entries', () => {
  render(
    <NavSection
      entries={mockEntries}
      sectionId="section1"
      activeItem="entry1"
    />,
  )

  expect(screen.getByRole('link', { name: 'Entry1' })).toBeInTheDocument()
  expect(screen.getByRole('link', { name: 'Entry2' })).toBeInTheDocument()
  expect(screen.getByRole('link', { name: 'Entry3' })).toBeInTheDocument()
})

it('properly renders the section title', () => {
  render(
    <NavSection
      entries={mockEntries}
      sectionId="section1"
      activeItem="entry1"
    />,
  )

  expect(screen.getByRole('button', { name: 'Section1' })).toBeInTheDocument()
})

it('marks the correct entry as active', () => {
  render(
    <NavSection
      entries={mockEntries}
      sectionId="section1"
      activeItem="entry1"
    />,
  )

  expect(screen.getByRole('link', { name: 'Entry1' })).toHaveClass(
    'pf-m-current',
  )
})

it('does not mark any entries as active if none are active', () => {
  render(
    <NavSection entries={mockEntries} sectionId="section1" activeItem="" />,
  )

  screen.getAllByRole('link').forEach((link) => {
    expect(link).not.toHaveClass('pf-m-current')
  })
})

it('matches snapshot', () => {
  const { asFragment } = render(
    <NavSection
      entries={mockEntries}
      sectionId="section1"
      activeItem="entry1"
    />,
  )
  expect(asFragment()).toMatchSnapshot()
})
