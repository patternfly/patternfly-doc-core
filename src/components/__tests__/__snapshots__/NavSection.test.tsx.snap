// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <li
    class="pf-v6-c-nav__item pf-m-expanded pf-m-current"
    data-ouia-component-id="OUIA-Generated-NavExpandable-8"
    data-ouia-component-type="PF6/NavExpandable"
    data-ouia-safe="true"
  >
    <button
      aria-expanded="true"
      class="pf-v6-c-nav__link"
      id="nav-section-section1"
    >
      Section1
      <span
        class="pf-v6-c-nav__toggle"
      >
        <span
          class="pf-v6-c-nav__toggle-icon"
        >
          <svg
            aria-hidden="true"
            class="pf-v6-svg"
            fill="currentColor"
            height="1em"
            role="img"
            viewBox="0 0 256 512"
            width="1em"
          >
            <path
              d="M224.3 273l-136 136c-9.4 9.4-24.6 9.4-33.9 0l-22.6-22.6c-9.4-9.4-9.4-24.6 0-33.9l96.4-96.4-96.4-96.4c-9.4-9.4-9.4-24.6 0-33.9L54.3 103c9.4-9.4 24.6-9.4 33.9 0l136 136c9.5 9.4 9.5 24.6.1 34z"
            />
          </svg>
        </span>
      </span>
    </button>
    <section
      aria-labelledby="nav-section-section1"
      class="pf-v6-c-nav__subnav"
    >
      <ul
        class="pf-v6-c-nav__list"
        role="list"
      >
        <li
          class="pf-v6-c-nav__item"
          data-ouia-component-id="OUIA-Generated-NavItem-22"
          data-ouia-component-type="PF6/NavItem"
          data-ouia-safe="true"
        >
          <a
            aria-current="page"
            class="pf-v6-c-nav__link pf-m-current"
            href="/section1/entry1"
            id="nav-entry-entry1"
          >
            <span
              class="pf-v6-c-nav__link-text"
            >
              Entry1
            </span>
          </a>
        </li>
        <li
          class="pf-v6-c-nav__item"
          data-ouia-component-id="OUIA-Generated-NavItem-23"
          data-ouia-component-type="PF6/NavItem"
          data-ouia-safe="true"
        >
          <a
            class="pf-v6-c-nav__link"
            href="/Section1/entry2"
            id="nav-entry-entry2"
          >
            <span
              class="pf-v6-c-nav__link-text"
            >
              Entry2
            </span>
          </a>
        </li>
        <li
          class="pf-v6-c-nav__item"
          data-ouia-component-id="OUIA-Generated-NavItem-24"
          data-ouia-component-type="PF6/NavItem"
          data-ouia-safe="true"
        >
          <a
            class="pf-v6-c-nav__link"
            href="/Section1/entry3"
            id="nav-entry-entry3"
          >
            <span
              class="pf-v6-c-nav__link-text"
            >
              Entry3
            </span>
          </a>
        </li>
      </ul>
    </section>
  </li>
</DocumentFragment>
`;
